src/main/java/duke/Ui.java

```diff
@@ -0,0 +1,144 @@
+package duke;
+
+import java.util.ArrayList;
+import java.util.concurrent.BlockingQueue;
+
+import duke.tasks.Task;
+
+/**
+ * Represents the user interface, responsible reading and writing messages.
+ */
+public class Ui {
+    private BlockingQueue<String> inputQueue;
+    private BlockingQueue<String> outputQueue;
+    private Formatter formatter = msg -> msg;
+    /**
+     * Constructor for a user interface.
+     */
+    public Ui(BlockingQueue<String> inputQueue, BlockingQueue<String> outputQueue) {
+        this.inputQueue = inputQueue;
+        this.outputQueue = outputQueue;
+    }
+
+    /**
+     * Use a formatter for outputting messages.
+     *
+     * @param f The formatter to use.
+     */
+    public void useFormatter(Formatter f) {
+        this.formatter = f;
+    }
+
+    /**
+     * Formats and writes a message to standard output.
+     *
+     * @param msg The message to format.
+     */
+    public void write(String msg) {
```

Comment:
Once again, you could consider more descriptive names like readMessage or writeMessage here.